rfcs/0051-emit-on-close-stream-executor.md

```diff
@@ -0,0 +1,136 @@
+SortBuffer consists of two parts, SortBufferCache and SortBufferTable.
+
+- SortBufferTable
+  - A stateTable to persist the buffered data.
+  - The primary key's first column is a watermark column. We call it as **sort key**
+- SortBufferCache
+  - A row cache on the SortBufferTable
+  - can accept watermark to know some rows have been complete
+  - can drain the complete row in the memory ordered by the sort key.
+
+And the SortBufferTable is just a normal StateTable, we will only need to implement the SortBufferCache.
+
+```rust
+impl SortBufferCache{
+  /// consume the next row ordered by the first column which is complete.
+  /// return None if there is no remaining complete row.
+  /// fill cache from the stateTable when cache miss.
+  /// will consume the row in memory.
+  fn next(&mut self, table: &StateTable) -> Option<Row>;
+
+  /// get the minimum sort key in rows which is complete.
+  /// return None if there is no remaining complete row.
+  /// fill cache from the stateTable when cache miss.
+  /// will not consume the row in memory.
+  fn peek_sort_key(&self, table: &StateTable) -> Option<ScalarImpl>;
+
+  /// use the first column's watermark to refresh the SortBuffer
+  /// can trigger more rows complete which can be returned by `self.next`
+  fn accept_watermark(&mut self, watermark: Watermark);
+  
+  /// apply changes operations
+  fn insert(&mut self, row: Row);
+  fn delete(&mut self, row: Row);
+  fn apply_chunk(&mut self, c: Chunk);
+
+  /// recovery from the stateTable, just gets the row belonging to stateTable's vnode.
+  fn recovery(&mut self, table: &StateTable);
+}
+
+```
+
+The component does not take over so many things. There are still lots of things that should be done by the executor, such as
+
+- all the operations should be applied on the table and cache, and the executor should write in duplicate.
+- after getting the complete data, the executor must delete the useless data by itself (delete one by one/range delete with the last key). Notice that it is different with state clean and the data must be deleted consistently.
+
+And then we will see why and how it works in different situations.
+
+## Executors
+
+### Sort
+
+  SortExecutor just has a executor which wraps a SortBuffer which can transform any stream into a append only stream ordered by a watermark column.
+
+- when stream chunk arrive
+  - write the chunk into stateTable
+  - write the chunk into SortBufferCache
+- when watermark arrive
+  - pass watermark into SortBufferCache
+- when barrier arrive
+  - drain complete rows from SortBufferCache, for each rows
+    - emit the row downstream
+    - delete the row in stateTable
+
+### SortAgg (with batch only agg calls)
+
+  Under "EMIT ON CLOSE" semantics, if there is watermark in group key, we can use SortExecutor for the input stream, and then we get a sorted stream which can be processed by "BatchSortAgg". Then, we can support all the aggregators which can run in batch mode.
```

Comment:
Another missing piece is the state. As we know the batch aggregate functions don't have support for states, as a result, IIUC, in order to reuse them in `SortAgg` you have to keep all these input rows in `SortAgg`'s state table and call the batch aggregate function at once to emit result. This may not be a good trade-off, especially for the value-state aggregate functions e.g. `count()`, `sum()`.

Reply:
No, actually, we do not need the aggregate functions' state here. Each group's data are totally materialized in the sort buffer table and when they are consumed in some epoch, it is a limited append only stream

Reply:
Yes, so we have another EMIT ON CLOSE GroupAgg implementation as described below. I think the SortAgg is just to solve some aggregation functions which does not have a streaming version.

Reply:
> I think the SortAgg is just to solve some aggregation functions which does not have a streaming version(hard to implement or with high cost)

I see. Makes sense.